src/deltalake_ext.rs

```diff
@@ -0,0 +1,322 @@
+        let storage = deltalake::get_backend_for_uri(table_path.as_str())?;
+
+        // Initialize an arrow schema ref from the delta table schema
+        let arrow_schema = <ArrowSchema as From<&Schema>>::from(schema);
+        let arrow_schema_ref = Arc::new(arrow_schema);
+
+        // Initialize writer properties for the underlying arrow writer
+        let writer_properties = WriterProperties::builder()
+            // TODO: Extract config/env for writer properties and set more than just compression
+            .set_compression(Compression::SNAPPY)
+            .build();
+
+        let cursor = InMemoryWriteableCursor::default();
+        // TODO: Handle error
+        let arrow_writer = ArrowWriter::try_new(cursor.clone(), arrow_schema_ref.clone(), Some(writer_properties.clone())).unwrap();
+
+        Ok(Self {
+            table_path,
+            storage,
+            arrow_schema_ref,
+            writer_properties,
+            cursor,
+            arrow_writer,
+            partition_values: HashMap::new(),
+            num_records: 0,
+        })
+    }
+
+    /// Writes the record batch in-memory and updates internal state accordingly.
+    /// This method buffers the write stream internally so it can be invoked for many record batches and flushed after the appropriate number of bytes has been written.
+    pub async fn write_record_batch(&mut self, partition_cols: &Vec<String>, record_batch: &RecordBatch) -> Result<(), DeltaWriterError> {
+        let partition_values = extract_partition_values(partition_cols, record_batch)?;
+
+        // TODO: Verify that this RecordBatch contains partition values that agree with previous partition values (if any) for the current writer context
+        self.partition_values = partition_values;
+
+        // TODO: collect/update column stats
+
+        // write the record batch to the held arrow writer
+        self.arrow_writer.write(record_batch)?;
+
+        // increment num records (for the entire file) by the number of records in the handled batch
+        self.num_records += record_batch.num_rows() as i64;
+
+        Ok(())
+    }
+
+    /// Returns the current byte length of the in memory buffer.
+    /// This should be used by the caller to decide when to call `complete_file`.
+    pub fn buffer_len(&self) -> Result<usize, DeltaWriterError> {
+        Ok(self.cursor.data().len())
+    }
+    
+    /// Writes the existing parquet bytes to storage and resets internal state to handle another round.
+    /// Returns an `Add` action representing the added file to be appended to the delta log.
+    pub async fn complete_file(&mut self, partition_cols: &Vec<String>) -> Result<Add, DeltaWriterError> {
+        let path = self.next_data_path(partition_cols, &self.partition_values).unwrap();
+
+        let obj_bytes = self.cursor.data();
+        let file_size = obj_bytes.len() as i64;
+
+        let storage_path = format!("{}/{}", self.table_path, path);
```

Comment:
should use the join path method from storage backend to make it work in all platforms.

Reply:
Nice suggestion. Fixed in https://github.com/delta-io/kafka-delta-ingest/pull/1/commits/c90ff3418c7bbf0e56cce7a7dda990e3c616140a#diff-5d8c9545386ed0b9dcf9dab1722a6b484e1d3014d9a2470788dd7b20509ca642R176.